src/enableDisableRules.ts

```diff
@@ -41,116 +41,114 @@ export class EnableDisableRulesWalker extends RuleWalker {
         }
     }
 
-    public visitSourceFile(node: ts.SourceFile) {
-        forEachComment(node, (fullText, _kind, pos) => {
-            return this.handlePossibleTslintSwitch(fullText.substring(pos.tokenStart, pos.end), node, pos);
+    public getEnableDisableRuleMap() {
+        forEachComment(this.sourceFile, (fullText, kind, pos) => {
+            const commentText = kind === ts.SyntaxKind.SingleLineCommentTrivia
+                ? fullText.substring(pos.tokenStart + 2, pos.end)
+                : fullText.substring(pos.tokenStart + 2, pos.end - 2);
+            return this.handleComment(commentText, pos);
         });
+
+        return this.enableDisableRuleMap;
     }
 
-    private getStartOfLinePosition(node: ts.SourceFile, position: number, lineOffset = 0) {
-        const line = ts.getLineAndCharacterOfPosition(node, position).line + lineOffset;
-        const lineStarts = node.getLineStarts();
+    private getStartOfLinePosition(position: number, lineOffset = 0) {
+        const line = ts.getLineAndCharacterOfPosition(this.sourceFile, position).line + lineOffset;
+        const lineStarts = this.sourceFile.getLineStarts();
         if (line >= lineStarts.length) {
             // next line ends with eof or there is no next line
-            return node.getFullWidth();
+            // undefined switches the rule until the end and avoids an extra array entry
+            return undefined;
         }
         return lineStarts[line];
     }
 
     private switchRuleState(ruleName: string, isEnabled: boolean, start: number, end?: number): void {
         const ruleStateMap = this.enableDisableRuleMap[ruleName];
+        if (ruleStateMap === undefined || // skip switches for unknown or disabled rules
+            isEnabled === ruleStateMap[ruleStateMap.length - 1].isEnabled // no need to add switch points if there is no change
+        ) {
+            return;
+        }
 
         ruleStateMap.push({
             isEnabled,
             position: start,
         });
 
         if (end) {
-            // switchRuleState method is only called when rule state changes therefore we can safely use opposite state
+            // we only get here when rule state changes therefore we can safely use opposite state
             ruleStateMap.push({
                 isEnabled: !isEnabled,
                 position: end,
             });
         }
     }
 
-    private getLatestRuleState(ruleName: string): boolean {
-        const ruleStateMap = this.enableDisableRuleMap[ruleName];
-
-        return ruleStateMap[ruleStateMap.length - 1].isEnabled;
+    private handleComment(commentText: string, pos: TokenPosition) {
+        const match = /^\s*tslint:/.exec(commentText);
```

Comment:
How about `/^\s*tslint:((enable|disable)(-(line|next-line))?):([\w-]*)/` and use the capture groups?